### PR TITLE
Catkin plugin: build with in-snap python.

### PR DESCRIPTION
**Pull request by kyrofa**
Our test cases all had python installed on the system, so the fact that the plugin used it went unnoticed. This PR fixes LP: [#1628990](https://bugs.launchpad.net/snapcraft/+bug/1628990) by using the in-snap python for build-time as well as run-time. It also pulls in gcc, libc-dev, and make as build packages.

### CODE REVIEW

File: snapcraft/plugins/catkin.py [hunk 1]
```diff
@@ -277,6 +283,22 @@ def build(self):
         self._finish_build()
 
     def _prepare_build(self):
+        # Fix all shebangs to use the in-snap python.
+        common.replace_in_file(self.rosdir, re.compile(r''),
+                               re.compile(r'#!.*python'),
```
Comment (josepht) [inline snapcraft/plugins/catkin.py]: should probably start with '^'

File: snapcraft/plugins/catkin.py [hunk 2]
```diff
@@ -277,6 +283,22 @@ def build(self):
...
+                               r'#!/usr/bin/env python')
+
+        # Also replace the python usage in 10.ros.sh to use the in-snap python.
+        ros10_file = os.path.join(self.rosdir,
+                                  'etc/catkin/profile.d/10.ros.sh')
+        if os.path.isfile(ros10_file):
```
Comment (josepht) [inline snapcraft/plugins/catkin.py]: I don't know if we care, but this is racey.  I think the SOP is to wrap the `open()` in a `try: ... except:`

File: snapcraft/plugins/catkin.py [hunk 3]
```diff
@@ -277,6 +283,22 @@ def build(self):
         self._finish_build()
 
     def _prepare_build(self):
+        # Fix all shebangs to use the in-snap python.
+        common.replace_in_file(self.rosdir, re.compile(r''),
```
Comment (sergiusens) [inline snapcraft/plugins/catkin.py]: use `fileutils` instead

File: snapcraft/plugins/catkin.py (hunk 1 above)
Comment (sergiusens) [inline snapcraft/plugins/catkin.py]: Agreed

File: snapcraft/plugins/catkin.py (hunk 2 above)
Comment (kyrofa) [inline snapcraft/plugins/catkin.py]: Of course we care! I blindly copied that from elsewhere. I've refactored a bit to fix both places.

File: snapcraft/plugins/catkin.py (hunk 1 above)
Comment (kyrofa) [inline snapcraft/plugins/catkin.py]: Indeed, done.

File: snapcraft/plugins/catkin.py (hunk 3 above)
Comment (kyrofa) [inline snapcraft/plugins/catkin.py]: Good catch, thanks! Done, and fixed one other place in this plugin as well.

File: snapcraft/plugins/catkin.py (hunk 2 above)
Comment (kyrofa) [inline snapcraft/plugins/catkin.py]: Ah, this was tricky. I missed the 'r+' on the mode-- we open for read/write. We need to know if it exists before we try, or we'll create it (and we don't want to if it's not already there). I've waffled back on this.

**Pull request merged**